tfx/orchestration/experimental/optimizations/beam_fusion.py

```diff
@@ -0,0 +1,346 @@
+    queue = [c for c in components if c in sources]
+    queue = sorted(queue, key=lambda c: c.id)
+
+    while queue:
+      component = queue.pop(0)
+      sorted_components.append(component)
+
+      new_sources = []
+      for child in component.downstream_nodes:
+        if child not in components:
+          continue
+
+        in_degrees[child] -= 1
+        if in_degrees[child] == 0:
+          new_sources.append(child)
+
+      # This extra sorting keeps a consistent topological ordering
+      new_sources = sorted(new_sources, key=lambda c: c.id)
+      queue.extend(new_sources)
+
+    return sorted_components
+
+  def _get_intersecting_subgraph(
+      self,
+      intersecting_component: base_node.BaseNode,
+      fuseable_subgraphs: List[List[base_node.BaseNode]]):
+
+    intersecting_subgraph = None
+    for subgraph in fuseable_subgraphs:
+      if intersecting_component in subgraph:
+        intersecting_subgraph = subgraph
+        break
+
+    return intersecting_subgraph
+
+  def _is_beam_component(self, component: base_node.BaseNode):
+    return issubclass(component.EXECUTOR_SPEC.executor_class,
+                      base_executor.FuseableBeamExecutor)
+
+  def _is_fuseable(self, child: base_node.BaseNode,
+                   current_subgraph: List[base_node.BaseNode]):
+    is_fuseable = True
+
+    # Conduct a BFS to ensure none of the child's ancestors (other than its'
+    # immediate parents and those parents' ancestors) are in current_subgraph.
+    queue = []
+    for parent in child.upstream_nodes:
+      if not parent in current_subgraph:
+        queue.append(parent)
+
+    while(is_fuseable and queue):
+      component = queue.pop(0)
+
+      for parent in component.upstream_nodes:
+        if parent in current_subgraph:
+          is_fuseable = False
+          break
+        queue.append(parent)
+
+    return is_fuseable
+
+  def _build_subgraph_from_source(
+      self,
+      source: base_node.BaseNode,
+      fuseable_subgraphs: List[List[base_node.BaseNode]],
+      visited: Set[base_node.BaseNode]):
+
+    subgraph = []
+
+    if source in visited:
+      return fuseable_subgraphs, visited
+
+    # Conduct a BFS on the candidate source.
+    queue = [source]
+    while queue:
+      component = queue.pop(0)
+
+      if component in visited:
+        continue
+
+      visited.add(component)
+      subgraph.append(component)
+
+      # Iterate through the children to find Beam components
+      for child in component.downstream_nodes:
+        if not self._is_beam_component(child):
+          continue
+
+        # Checks if the child is in an explored subgraph that needs to be
+        # fused into the current subrgaph
+        if child in visited:
+          intersecting_subgraph = self._get_intersecting_subgraph(
+              child, fuseable_subgraphs)
+          if intersecting_subgraph:
+            subgraph.extend(intersecting_subgraph)
+            fuseable_subgraphs.remove(intersecting_subgraph)
```

Comment:
This is a good edge case. I think that the two subgraphs that will be returned will be `[A, B]` and `[E, F]`, since `D` has parents with ancestors that are in a subgraph. In other words, `D` will never be added to either subgraph because `C*` is its parent. Let me know if you want me to write up a supporting testcase in `beam_fusion_test.py`.